src/pathfinding_service/model/channel_view.py

```diff
@@ -62,13 +64,40 @@ def update_capacity(
         if reveal_timeout is not None:
             self.reveal_timeout = reveal_timeout
 
-    def forward_fee_sender(self, amount: PaymentWithFeeAmount) -> FeeAmount:
-        """Return the mediation fee for this channel when transferring the given amount"""
-        return self.fee_schedule_sender.fee_payer(amount, Balance(self.capacity))
+    def backwards_fee_sender(
+        self, balance: Balance, amount: PaymentWithFeeAmount
+    ) -> Optional[FeeAmount]:
+        """Returns the mediation fee for this channel when transferring the given amount"""
+        try:
+            return self.fee_schedule_sender.fee_payer(amount, balance)
+        except UndefinedMediationFee:
+            return None
 
-    def forward_fee_receiver(self, amount: PaymentWithFeeAmount) -> FeeAmount:
-        """Return the mediation fee for this channel when receiving the given amount"""
-        return self.fee_schedule_receiver.fee_payee(amount, Balance(self.capacity))
+    def backwards_fee_receiver(
+        self, balance: Balance, amount: PaymentWithFeeAmount
+    ) -> Optional[FeeAmount]:
+        """Returns the mediation fee for this channel when receiving the given amount"""
+
+        def fee_in(imbalance_fee: FeeAmount) -> FeeAmount:
+            return FeeAmount(
+                round(
+                    (
+                        (amount + self.fee_schedule_receiver.flat + imbalance_fee)
+                        / (1 - self.fee_schedule_receiver.proportional / 1e6)
+                    )
+                    - amount
+                )
+            )
+
+        try:
+            imbalance_fee = self.fee_schedule_receiver.imbalance_fee(
+                amount=PaymentWithFeeAmount(amount - fee_in(imbalance_fee=FeeAmount(0))),
```

Comment:
The sign has to be flipped on this one:
```suggestion
                amount=PaymentWithFeeAmount(-(amount - fee_in(imbalance_fee=FeeAmount(0)))),
```